### ISSUE TITLE
403 error while passing token to directline

### ISSUE THREAD

**Issue by rambabuk1222** [microsoft/BotFramework-WebChat#2607]
## Screenshots

<%@ Page Title="Home Page" Language="C#" MasterPageFile="~/Site.Master" AutoEventWireup="true" CodeBehind="Default.aspx.cs" Inherits="BotFrameWebChat._Default" %>

<asp:Content ID="BodyContent" ContentPlaceHolderID="MainContent" runat="server">
     <script src="https://cdn.botframework.com/botframework-webchat/latest/webchat.js"></script>
    <style>
      html,
      body {
        height: 100%;
      }
      body {
        margin: 0;
      }

      #webchat {
        height: 100%;
        width: 100%;
      }
    </style>
    <div class="jumbotron">
        <h1>ASP.NET</h1>
        <p class="lead">ASP.NET is a free web framework for building great Web sites and Web applications using HTML, CSS, and JavaScript.</p>
        <p><a href="http://www.asp.net" class="btn btn-primary btn-lg">Learn more &raquo;</a></p>
    </div>
    
    <div id="webchat" role="main"></div>
    <script>
         function createFetchSpeechServicesCredentials() {
        //let expireAfter = 0;
        //let lastResult = {};
             debugger;
        return async () => {
          // Fetch a new token if the existing one is expiring.
          // The following article mentioned the token is only valid for 10 minutes.
          // We will invalidate the token after 5 minutes.
          // https://docs.microsoft.com/en-us/azure/cognitive-services/authentication#authenticate-with-an-authentication-token
          if (Date.now() > expireAfter) {
            var speechServicesTokenRes = await fetch(
              'http://localhost:4000/tokengenerator',
              { method: 'GET' }
            );

              lastResult = await speechServicesTokenRes.json();
              debugger;
            expireAfter = Date.now() + 300000;
          }

          return lastResult;
        };
      }
        var fetchSpeechServicesCredentials = createFetchSpeechServicesCredentials();
        async function fetchSpeechServicesRegion() {
        return (await fetchSpeechServicesCredentials()).region;
      }

      async function fetchSpeechServicesToken() {
        return (await fetchSpeechServicesCredentials()).token;
      }
      (async function() {
        // In this demo, we are using Direct Line token from MockBot.
        // Your client code must provide either a secret or a token to talk to your bot.
        // Tokens are more secure. To learn about the differences between secrets and tokens
        // and to understand the risks associated with using secrets, visit https://docs.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-direct-line-3-0-authentication?view=azure-bot-service-4.0
          debugger;
        var res = await fetch('http://localhost:4000/tokengenerator', { method: 'GET' });
        var { token } = await res.json();

           const directLineTokenResponse = await fetch('http://localhost:4000/tokengenerator', {
          method: 'GET'
           });
          debugger;
        var { token } = await directLineTokenResponse.body();
          debugger;
        //// Create the ponyfill factory function, which can be called to create a concrete implementation of the ponyfill.
        var webSpeechPonyfillFactory = await window.WebChat.createCognitiveServicesSpeechServicesPonyfillFactory({
          // We are passing the Promise function to the authorizationToken field.
          // This function will be called every time the token is being used.
          authorizationToken: fetchSpeechServicesToken,

          // In contrast, we are only fetching the region once.
          region: await fetchSpeechServicesRegion()
        });
         
         window.WebChat.renderWebChat(
          {
                 directLine: window.WebChat.createDirectLine({ token }),
            selectVoice: (voices, activity) =>
              // If the activity is in zh-HK, use a voice with keyword "TracyRUS" (Cantonese).
              // Otherwise, use "JessaNeural" (preferred) or "Jessa".
              activity.locale === 'zh-HK'
                ? voices.find(({ name }) => /TracyRUS/iu.test(name))
                : voices.find(({ name }) => /JessaNeural/iu.test(name)) ||
                  voices.find(({ name }) => /Jessa/iu.test(name)),
            webSpeechPonyfillFactory
          },
          document.getElementById('webchat')
        );

        document.querySelector('#webchat > *').focus();
      })().catch(err => console.error(err));
    </script>
</asp:Content>
screenshots -->

## Version
https://cdn.botframework.com/botframework-webchat/latest/webchat.js

## Describe the bug

My local site generating token correctly . After passing generated token to directline, We are getting 403 bad request error below line
directLine: window.WebChat.createDirectLine({ token }),

## Expected behavior

After passing token to directline method, We should see  chat mic enabled and working fine.

**Comment by stevkan** [comment]
I tried your code and with a few minor adjustments it works fine for me.

First, I uncommented the `expireAfter` and `lastResult` variables as that was generating an error without being defined:

```js
  let expireAfter = 0;
  let lastResult = {};
```

Next, I'm not sure why you are acquiring a token twice. I commented out the `directLineTokenResponse` fetch and associated code. However, if this code is necessary, then you need to change

```js
var { token } = await directLineTokenResponse.body();
```
to 

```js
var { token } = await directLineTokenResponse.json();
```

After these changes, it worked fine for me.

If the issue persists, I'd be curious to see the code you use for generating tokens via the `tokengenerator` API call. I noticed you use the same API for getting a token for both speech services and direct line. Is this intentional? The direct line token API (`v3/directline/tokens/generate`) can't be used for authenticating speech services.

**Comment by rambabuk1222** [comment]
Thanks for your response. I will make changes as you suggested.

**Comment by rambabuk1222** [comment]
I have tried your suggestions. This time webchat is able to get token and region from my api and I saw microphone icon is enabled. When I try to speak it is giving some javascript errors as shown in the screenshot below.

![image](https://user-images.githubusercontent.com/12018103/70013661-4e283600-1570-11ea-8678-38ae69d19129.png)


![image](https://user-images.githubusercontent.com/12018103/70013553-fee20580-156f-11ea-913a-5ea1817b5fd9.png)

![image](https://user-images.githubusercontent.com/12018103/70013632-3650b200-1570-11ea-9181-9192fcd5fd8d.png)

Thanks,
Ram
